ra/registration-authority.go

```diff
@@ -452,9 +452,16 @@ func (ra *RegistrationAuthorityImpl) MatchesCSR(cert core.Certificate, csr *x509
 func (ra *RegistrationAuthorityImpl) checkAuthorizations(names []string, registration *core.Registration) error {
 	now := ra.clk.Now()
 	var badNames []string
+	for i := range names {
+		names[i] = strings.ToLower(names[i])
+	}
+	auths, err := ra.SA.GetValidAuthorizations(registration.ID, names, now)
+	if err != nil {
+		return err
+	}
 	for _, name := range names {
-		authz, err := ra.SA.GetLatestValidAuthorization(registration.ID, core.AcmeIdentifier{Type: core.IdentifierDNS, Value: name})
-		if err != nil || authz.Expires.Before(now) {
+		authz := auths[name]
```

Comment:
That looks good. Let's add a little paranoia and add a check that authz.Expires is not nil, since we'll panic otherwise. I'm thinking:

```
if authz == nil {
  badNames = append(badNames, name)
} else if authz.Expires == nil {
  return fmt.Errorf("Found an authorization with a nil Expires field")
} else if authz.Expires.Before(now) {
  badNames = append(badNames, name)
}
```